File: netguard.py
#!/usr/bin/env python3
"""Network hardening helpers: loopback-only URLs, probe targets, TTL cache.

SSRF posture for this console: outbound operator URLs (LAPI, metrics, Traefik API)
must resolve to loopback. PUBLIC_IP probe targets must be literal IP addresses
(no DNS). Probe Host/SNI values must be hostname-safe (no header injection).
"""
from __future__ import annotations

import ipaddress
import re
import threading
import time
from typing import Any
from urllib.parse import urlparse

_HOST_RE = re.compile(r"^[A-Za-z0-9](?:[A-Za-z0-9.-]{0,251}[A-Za-z0-9])?$")
_LOOPBACK_NAMES = frozenset({"127.0.0.1", "localhost", "::1"})


class TtlCache:
    """Thread-safe TTL cache for expensive local probes/scrapes."""

    def __init__(self) -> None:
        self._data: dict[Any, tuple[Any, float]] = {}
        self._lock = threading.Lock()

    def get(self, key: Any) -> Any | None:
        now = time.monotonic()
        with self._lock:
            item = self._data.get(key)
            if item is None:
                return None
            value, exp = item
            if now >= exp:
                del self._data[key]
                return None
            return value

    def set(self, key: Any, value: Any, ttl: float) -> None:
        if ttl <= 0:
            return
        with self._lock:
            self._data[key] = (value, time.monotonic() + ttl)

    def clear(self) -> None:
        with self._lock:
            self._data.clear()


def sanitize_hostname(host: str) -> str:
    """Reject CR/LF/null and non-hostname characters (SNI / Host header)."""
    value = (host or "").strip().lower()
    if not value or len(value) > 253:
        raise ValueError("invalid host")
    if any(ch in value for ch in ("\r", "\n", "\x00", "/", "\\", " ", ":", "@")):
        raise ValueError("invalid host")
    if not _HOST_RE.match(value):
        raise ValueError("invalid host")
    return value


def assert_probe_server(value: str, *, name: str = "PUBLIC_IP") -> str:
    """PUBLIC_IP must be empty or a literal IPv4/IPv6 address (no DNS)."""
    raw = (value or "").strip()
    if not raw:
        return ""
    try:
        return str(ipaddress.ip_address(raw))
    except ValueError as exc:
        raise ValueError(f"{name} must be a literal IP address") from exc


def assert_loopback_http_url(url: str, *, name: str = "url") -> str:
    """Allow only http(s) to loopback hostnames/addresses."""
    raw = (url or "").strip()
    if not raw:
        raise ValueError(f"{name} is empty")
    parsed = urlparse(raw)
    if parsed.scheme not in ("http", "https"):
        raise ValueError(f"{name} scheme must be http or https")
    if parsed.username or parsed.password:
        raise ValueError(f"{name} must not embed credentials")
    host = (parsed.hostname or "").lower()
    if not host:
        raise ValueError(f"{name} missing host")
    if host in _LOOPBACK_NAMES:
        return raw.rstrip("/")
    try:
        if ipaddress.ip_address(host).is_loopback:
            return raw.rstrip("/")
    except ValueError:
        pass
    raise ValueError(f"{name} must target loopback (got {host})")


def env_int(name: str, default: int, *, minimum: int = 0, maximum: int = 10_000) -> int:
    import os

    raw = os.environ.get(name, "")
    if not str(raw).strip():
        return default
    try:
        value = int(raw)
    except ValueError:
        return default
    return max(minimum, min(maximum, value))


def read_limited(resp, max_bytes: int = 2_000_000) -> bytes:
    """Bound urllib response bodies to avoid memory exhaustion from local services."""
    chunks: list[bytes] = []
    total = 0
    while True:
        chunk = resp.read(65536)
        if not chunk:
            break
        total += len(chunk)
        if total > max_bytes:
            raise RuntimeError("response exceeds size limit")
        chunks.append(chunk)
    return b"".join(chunks)


class RateLimiter:
    """Sliding-window rate limiter (per-key, in-process)."""

    def __init__(self, limit: int, window_sec: float) -> None:
        self.limit = max(1, int(limit))
        self.window = max(0.5, float(window_sec))
        self._hits: dict[str, list[float]] = {}
        self._lock = threading.Lock()

    def allow(self, key: str) -> bool:
        now = time.monotonic()
        cutoff = now - self.window
        with self._lock:
            bucket = [t for t in self._hits.get(key, []) if t > cutoff]
            if len(bucket) >= self.limit:
                self._hits[key] = bucket
                return False
            bucket.append(now)
            self._hits[key] = bucket
            # opportunistic prune of idle keys
            if len(self._hits) > 4096:
                stale = [k for k, v in self._hits.items() if not v or v[-1] < cutoff]
                for k in stale[:512]:
                    self._hits.pop(k, None)
            return True

File: rust/src/netguard.rs
//! Network hardening helpers (port of netguard.py).

use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

use anyhow::{anyhow, Result};
use regex::Regex;
use serde_json::Value;

static HOST_RE: OnceLock<Regex> = OnceLock::new();

fn host_re() -> &'static Regex {
    HOST_RE.get_or_init(|| {
        Regex::new(r"^[A-Za-z0-9](?:[A-Za-z0-9.-]{0,251}[A-Za-z0-9])?$").expect("HOST_RE")
    })
}

pub fn sanitize_hostname(host: &str) -> Result<String> {
    let value = host.trim().to_lowercase();
    if value.is_empty() || value.len() > 253 {
        return Err(anyhow!("invalid host"));
    }
    if value.contains(['\r', '\n', '\0', '/', '\\', ' ', ':', '@']) {
        return Err(anyhow!("invalid host"));
    }
    if !host_re().is_match(&value) {
        return Err(anyhow!("invalid host"));
    }
    Ok(value)
}

pub fn assert_probe_server(value: &str, name: &str) -> Result<String> {
    let raw = value.trim();
    if raw.is_empty() {
        return Ok(String::new());
    }
    raw.parse::<IpAddr>()
        .map(|ip| ip.to_string())
        .map_err(|e| anyhow!("{name} must be a literal IP address: {e}"))
}

pub fn assert_loopback_http_url(url: &str, name: &str) -> Result<String> {
    let raw = url.trim();
    if raw.is_empty() {
        return Err(anyhow!("{name} is empty"));
    }
    let parsed =
        url::Url::parse(raw).map_err(|_| anyhow!("{name} scheme must be http or https"))?;
    if parsed.scheme() != "http" && parsed.scheme() != "https" {
        return Err(anyhow!("{name} scheme must be http or https"));
    }
    if parsed.username() != "" || parsed.password().is_some() {
        return Err(anyhow!("{name} must not embed credentials"));
    }
    let host = parsed.host_str().unwrap_or("").to_lowercase();
    if host.is_empty() {
        return Err(anyhow!("{name} missing host"));
    }
    if matches!(host.as_str(), "127.0.0.1" | "localhost" | "::1") {
        return Ok(raw.trim_end_matches('/').to_string());
    }
    if let Ok(ip) = host.parse::<IpAddr>() {
        if ip.is_loopback() {
            return Ok(raw.trim_end_matches('/').to_string());
        }
    }
    Err(anyhow!("{name} must target loopback (got {host})"))
}

pub fn env_int(name: &str, default: i64, minimum: i64, maximum: i64) -> i64 {
    let raw = std::env::var(name).unwrap_or_default();
    if raw.trim().is_empty() {
        return default;
    }
    match raw.parse::<i64>() {
        Ok(v) => v.clamp(minimum, maximum),
        Err(_) => default,
    }
}

pub struct TtlCache {
    inner: Mutex<HashMap<String, (Value, Instant)>>,
}

impl TtlCache {
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(HashMap::new()),
        }
    }

    pub fn get(&self, key: &str) -> Option<Value> {
        let now = Instant::now();
        let mut guard = self.inner.lock().ok()?;
        if let Some((value, exp)) = guard.get(key) {
            if now < *exp {
                return Some(value.clone());
            }
            guard.remove(key);
        }
        None
    }

    pub fn set(&self, key: &str, value: Value, ttl: f64) {
        if ttl <= 0.0 {
            return;
        }
        if let Ok(mut guard) = self.inner.lock() {
            guard.insert(
                key.to_string(),
                (value, Instant::now() + Duration::from_secs_f64(ttl)),
            );
        }
    }

    pub fn clear(&self) {
        if let Ok(mut guard) = self.inner.lock() {
            guard.clear();
        }
    }
}

impl Default for TtlCache {
    fn default() -> Self {
        Self::new()
    }
}

pub struct RateLimiter {
    limit: usize,
    window: Duration,
    hits: Mutex<HashMap<String, Vec<Instant>>>,
}

impl RateLimiter {
    pub fn new(limit: usize, window_sec: f64) -> Self {
        Self {
            limit: limit.max(1),
            window: Duration::from_secs_f64(window_sec.max(0.5)),
            hits: Mutex::new(HashMap::new()),
        }
    }

    pub fn allow(&self, key: &str) -> bool {
        let now = Instant::now();
        let cutoff = now - self.window;
        let Ok(mut guard) = self.hits.lock() else {
            return true;
        };
        let bucket = guard.entry(key.to_string()).or_default();
        bucket.retain(|t| *t > cutoff);
        if bucket.len() >= self.limit {
            return false;
        }
        bucket.push(now);
        if guard.len() > 4096 {
            let stale: Vec<String> = guard
                .iter()
                .filter(|(_, v)| v.is_empty() || v.last().is_some_and(|t| *t < cutoff))
                .map(|(k, _)| k.clone())
                .take(512)
                .collect();
            for k in stale {
                guard.remove(&k);
            }
        }
        true
    }
}
